File: api/Wrappers/ApiResponse.cs
namespace DentalCore.API.Wrappers;

/// <summary>
/// Envelope padronizado para todas as respostas da API.
/// Garantindo que o cliente frontend sempre receba um formato conhecido.
/// </summary>
public class ApiResponse<T>
{
    public bool Sucesso { get; set; }
    public T? Dados { get; set; }
    public string? Mensagem { get; set; }

    public ApiResponse()
    {
    }

    public ApiResponse(T dados, string mensagem = "")
    {
        Sucesso = true;
        Dados = dados;
        Mensagem = mensagem;
    }

    public ApiResponse(string mensagem)
    {
        Sucesso = false;
        Mensagem = mensagem;
    }

    public static ApiResponse<T> Ok(T dados, string mensagem = "") => new(dados, mensagem);
    public static ApiResponse<T> Erro(string mensagem) => new(mensagem);
}

File: MobileApp/lib/models/api_response.dart
/// Espelho exato do `ApiResponse<T>` do backend C#.
/// Envelope padrão de TODAS as respostas da API.
class ApiResponse<T> {
  final bool sucesso;
  final T? dados;
  final String? mensagem;

  ApiResponse({
    required this.sucesso,
    this.dados,
    this.mensagem,
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic) fromJsonT,
  ) {
    return ApiResponse<T>(
      sucesso:  json['sucesso'] ?? false,
      mensagem: json['mensagem'],
      dados: json['dados'] != null ? fromJsonT(json['dados']) : null,
    );
  }

  factory ApiResponse.erro(String mensagem) {
    return ApiResponse<T>(sucesso: false, mensagem: mensagem);
  }
}
